libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -0,0 +1,252 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/analysis/ImmutableValidator.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace solidity::frontend;
+
+void ImmutableValidator::analyze()
+{
+	m_inConstructionContext = true;
+
+	for (VariableDeclaration const* stateVar: m_currentContract.stateVariablesIncludingInherited())
+		if (stateVar->value())
+		{
+			stateVar->value()->accept(*this);
+			solAssert(m_initializedStateVariables.insert(stateVar).second, "");
+		}
+
+	for (auto const* contract: m_currentContract.annotation().linearizedBaseContracts  | boost::adaptors::reversed)
+	{
+		m_inConstructionContext = true;
+
+		if (contract->constructor())
+		{
+			contract->constructor()->accept(*this);
+			m_visitedCallables.insert(contract->constructor());
+		}
+
+		for (std::shared_ptr<InheritanceSpecifier> const inheritSpec: contract->baseContracts())
+			if (auto args = inheritSpec->arguments())
+				ASTNode::listAccept(*args, *this);
+
+		m_inConstructionContext = false;
```

Comment:
Is it essential that you first visit all functions of one contract instead of visiting all constructors first and then everything else?

Reply:
I think it would be cleaner to first visit everything that can be called from construction context and mark it as already visited and only then visit everything else.